apps/src/applab/designMode.js

```diff
@@ -353,40 +349,78 @@ designMode.toggleDesignMode = function(enable) {
+    var wrapper = elm.wrap('<div>').parent().resizable({
+      alsoResize: elm,
+      stop: function () {
+        Applab.levelHtml = designMode.serializeToLevelHtml();
+      }
+    }).draggable({
+      cancel: false,  // allow buttons and inputs to be dragged
+      drag: function (event, ui) {
+        // draggables are not compatible with CSS transform-scale,
+        // so adjust the position in various ways here.
+
+        // dragging
+        var div = document.getElementById('divApplab');
+        var xScale = div.getBoundingClientRect().width / div.offsetWidth;
+        var yScale = div.getBoundingClientRect().height / div.offsetHeight;
+        var changeLeft = ui.position.left - ui.originalPosition.left;
+        var newLeft  = (ui.originalPosition.left + changeLeft) / xScale;
+        var changeTop = ui.position.top - ui.originalPosition.top;
+        var newTop = (ui.originalPosition.top + changeTop) / yScale;
+
+        // snap top-left corner to nearest location in the grid
+        newLeft -= (newLeft + GRID_SIZE / 2) % GRID_SIZE - GRID_SIZE / 2;
+        newTop -= (newTop + GRID_SIZE / 2) % GRID_SIZE - GRID_SIZE / 2;
+
+        // containment
+        var container = $('#divApplab');
+        var maxLeft = container.outerWidth() - ui.helper.outerWidth(true);
+        var maxTop = container.outerHeight() - ui.helper.outerHeight(true);
+        newLeft = Math.min(newLeft, maxLeft);
+        newLeft = Math.max(newLeft, 0);
+        newTop = Math.min(newTop, maxTop);
+        newTop = Math.max(newTop, 0);
+
+        ui.position.left = newLeft;
+        ui.position.top = newTop;
+      },
+      stop: function () {
+        Applab.levelHtml = designMode.serializeToLevelHtml();
+      }
+    }).css('position', 'absolute');
 
-      // snap top-left corner to nearest location in the grid
-      newLeft -= (newLeft + GRID_SIZE / 2) % GRID_SIZE - GRID_SIZE / 2;
-      newTop -= (newTop + GRID_SIZE / 2) % GRID_SIZE - GRID_SIZE / 2;
-
-      // containment
-      var container = $('#divApplab');
-      var maxLeft = container.outerWidth() - ui.helper.outerWidth(true);
-      var maxTop = container.outerHeight() - ui.helper.outerHeight(true);
-      newLeft = Math.min(newLeft, maxLeft);
-      newLeft = Math.max(newLeft, 0);
-      newTop = Math.min(newTop, maxTop);
-      newTop = Math.max(newTop, 0);
-
-      ui.position.left = newLeft;
-      ui.position.top = newTop;
-    },
-    stop: function(event, ui) {
-      Applab.levelHtml = designMode.serializeToLevelHtml();
-    }
+    wrapper.css({
+      top: elm.css('top'),
+      left: elm.css('left')
+    });
+
+    elm.css({
+      position: ''
+    })
+  });
+}
+
+/**
+ * Inverse of `makeDraggable`.
+ * @param {jQuery} jq jQuery object containing DOM elements to make undraggable.
+ */
+function makeUndraggable(jq) {
+  jq.each(function () {
+    var wrapper = $(this);
+    var elm = $(':first-child', wrapper);
+
+    wrapper.resizable('destroy').draggable('destroy');
```

Comment:
I think we need to check that the element is draggable and resizable before doing these, otherwise the call to 'destroy' may fail. For example, if we are in "run" mode while in design mode the elements are not draggable. Alternately, if it is easy to explain why this will never happen, then some comments explaining why would be helpful.